### DISCUSSION TITLE
Threading issue when using yolov8 on Jetson Orin NX

### DISCUSSION THREAD

**Discussion by Hokaage** [ultralytics/ultralytics#10056] [Q&A]
When using Jetson I want to render two different images or camera images using threading but only one black window opens. When I run the same code on my computer it works correctly.

```
from ultralytics import YOLO
import threading

def predict_test():
    model = YOLO('yolov8n.pt')

    results = model(source="video.mp4", stream=True, show=True)

    for result in results:
        boxes = result.boxes
        masks = result.masks
        keypoints = result.keypoints
        probs = result.probs

def predict_test2():
    model = YOLO('yolov8n.pt')

    results = model(source="video2.mp4", stream=True, show=True)

    for result in results:
        boxes = result.boxes
        masks = result.masks
        keypoints = result.keypoints
        probs = result.probs


thread1 = threading.Thread(target=predict_test)
thread2 = threading.Thread(target=predict_test2)

thread1.start()
thread2.start()

thread1.join()
thread2.join()

```
I am new to these issues and my English is not very good, please answer accordingly.

**Comment by pderrenger** [comment]
Hello there! 😊

Your issue with threading when using YOLOv8 on Jetson Orin NX sounds like it might be related to how OpenCV's `imshow` function behaves in threaded environments, especially on platforms like Jetson. Each thread might be attempting to open its window on the same thread, causing conflicts. 

One possible workaround is to use a queue to collect results from each thread and then display them in the main thread. Here's a simplified example:

```python
from ultralytics import YOLO
import threading
from queue import Queue
import cv2

# Create a queue to hold inference results
results_queue = Queue()

def predict_test(source, results_queue):
    model = YOLO('yolov8n.pt')
    results = model(source=source, stream=True)
    for result in results:
        # Instead of showing the results here, put them into the queue
        results_queue.put(result.plot())

# Thread target functions now take 'source' and 'results_queue' as parameters
thread1 = threading.Thread(target=predict_test, args=("video.mp4", results_queue))
thread2 = threading.Thread(target=predict_test, args=("video2.mp4", results_queue))

thread1.start()
thread2.start()

# In the main thread, retrieve and display the results
while not results_queue.empty() or thread1.is_alive() or thread2.is_alive():
    if not results_queue.empty():
        result_img = results_queue.get()
        cv2.imshow("YOLOv8 Inference", result_img)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

cv2.destroyAllWindows()
thread1.join()
thread2.join()
```

This approach ensures that `imshow` and related display operations happen on the main thread, potentially avoiding the threading issue you encountered.

Let us know how it goes, and feel free to ask if you have more questions!

**Comment by Hokaage** [comment]
First of all, thank you for your reply. This time only one window opened and the frames kept moving from one video to another. Is there no way to play these videos in two different windows?

Addition: I’m not sure if I solved it correctly, but I found a way to run it in two different windows by changing the code like this:

```
from ultralytics import YOLO
import threading
from queue import Queue
import cv2
import time

# Create two queues to hold inference results
results_queue1 = Queue()
results_queue2 = Queue()

# Create a flag to control the threads' execution
stop_threads = False

def predict_test(source, results_queue, window_name):
    global stop_threads
    model = YOLO('yolov8n.pt')
    while not stop_threads:
        results = model(source=source, stream=True)
        for result in results:
            # Check the flag again before putting the result into the queue
            if stop_threads:
                break
            results_queue.put(result.plot())
        time.sleep(0.1)  # Add a short delay to prevent overloading the CPU

# Thread target functions now take 'source', 'results_queue', and 'window_name' as parameters
thread1 = threading.Thread(target=predict_test, args=("video.mp4", results_queue1, "Pencere 1"))
thread2 = threading.Thread(target=predict_test, args=("video.mp4", results_queue2, "Pencere 2"))

thread1.start()
thread2.start()

# In the main thread, retrieve and display the results from both queues
while True:
    if not results_queue1.empty():
        result_img1 = results_queue1.get()
        cv2.imshow("Pencere 1", result_img1)
    if not results_queue2.empty():
        result_img2 = results_queue2.get()
        cv2.imshow("Pencere 2", result_img2)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        stop_threads = True
        break

cv2.destroyAllWindows()

# Wait for both threads to finish
thread1.join()
thread2.join()

```

But this time, when I run the code for the first time in jetson it works, but when I stop and run it again it doesn't give any output, it just stops.

**Reply by pderrenger** [reply]
@Hokaage thank you for sharing your solution! It looks like you've come up with an inventive approach to run inferences in parallel and display them in separate windows. Using queues to manage the results from each thread is a smart way to handle the concurrent output. 👍

For the issue where the code continues to run in the terminal even after closing the windows, you might consider adding a shared variable (like an `Event` from the `threading` module) that your threading functions check to see if they should exit. Here's a small modification to your code:

```python
from threading import Event

# Create a shared Event for threads to check
shutdown_event = Event()

# Modify your thread worker functions to periodically check this event
def predict_test(source, results_queue, window_name):
    model = YOLO('yolov8n.pt')
    results = model(source=source, stream=True)
    for result in results:
        if shutdown_event.is_set():  # Check if shutdown event is set
            break
        results_queue.put(result.plot())

# Update your main loop to set the event when 'q' is pressed
if cv2.waitKey(1) & 0xFF == ord('q'):
    shutdown_event.set()  # Signal all threads to shutdown
```

This approach allows your threads to cleanly exit when the user presses 'q'. Don't forget to start and join your threads as you did in your original code. This should help ensure that when you close the screens and stop the program, it exits entirely without leaving any output in the terminal.

Keep up the great work! If you have any more questions or need further assistance, feel free to ask.
